### PR TITLE
DevOps: Add bug creation logic to ADO sync job

### PR DESCRIPTION
**Pull request by and-rewsmith**
Last PR in the series for the bug automation. With these changes, the service is able to create bugs for all failing builds. Specifically it works like this:
1. Scrape builds from devops rest api
2. Filter the scraped builds and keep only those that occurred in the last hour. Then filter out all the builds that we already logged bugs for.
3. Open bugs for the filtered builds.
4. Update the db to track which bugs we have opened for which builds.

This process allows us to open bugs for every failing build starting from when the service first starts. The service won't create tons of bugs at once since it only will do so for builds in the last hour. This means, if necessary, the db tables can be dropped and repopulated safely without many duplicates occurring.

When the service creates a bug it will title it with this format:
```
Test failure on refs/heads/release/1.2: EndToEndTest 20210623.2
```

In the description will be this message:
> This bug is autogenerated by the vsts-pipeline-sync service. Link to failing build:
<LINK_REDACTED>

Then we add a hyperlink to the bug hyperlink sections.

### CODE REVIEW

File: tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs [hunk 1]
```diff
@@ -60,6 +62,61 @@ public async Task RunAsync(TimeSpan waitPeriodAfterEachUpdate, CancellationToken
...
+                        Console.WriteLine(e.Message);
+                        Console.WriteLine("Create bug failed. Will retry later.");
+                    }
+                }
+            }
+
+            if (buildIdToBugId.Count == 0)
+            {
+                return;
+            }
```
Comment (and-rewsmith) [inline tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs:94]: There might be a cleaner way to do this.

File: tools/IoTEdgeDevOps/DevOpsLib/BugWiqlManagement.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+            if (!result.ContainsKey("queryType"))
+            {
+                return 0;
+            }
+
+            return result["workItems"].Count();
+        }
+    }
+}
```
Comment (and-rewsmith) [inline tools/IoTEdgeDevOps/DevOpsLib/BugWiqlManagement.cs:62]: The diff here is a bit misleading. This file has no changes. I renamed `BugManagement.cs` -> `BugWiqlManagement.cs` for clarity. Then I made a new file for creating the bugs called `BugManagement.cs`.

File: tools/IoTEdgeDevOps/DevOpsLib/BugManagement.cs [hunk 3]
```diff
@@ -26,37 +26,75 @@ public BugManagement(DevOpsAccessSetting accessSetting)
...
                 Console.WriteLine(e.Call.RequestBody);
                 Console.WriteLine(e.Call.Response.StatusCode);
                 Console.WriteLine(e.Call.Response.ResponseMessage);
-                return 0;
-            }
 
-            if (!result.ContainsKey("queryType"))
-            {
-                return 0;
+                throw new Exception(message);
```
Comment (and-rewsmith) [inline tools/IoTEdgeDevOps/DevOpsLib/BugManagement.cs:94]: I am rethrowing as I want this func to make it clear that something has gone wrong. The only success case for this function call should be if a bug is successfully made.

File: tools/IoTEdgeDevOps/DevOpsLib/BugManagement.cs [hunk 4]
```diff
@@ -2,15 +2,15 @@
 namespace DevOpsLib
```
Comment (and-rewsmith) [inline tools/IoTEdgeDevOps/DevOpsLib/BugManagement.cs:2]: The diff here is a bit misleading. I renamed BugManagement.cs -> BugWiqlManagement.cs for clarity. Then I made a new file for creating the bugs called BugManagement.cs.

The diff here is between the wiql api file (left), which has not been changed but just moved, and the new file (right) which took on the same name as the wiql api file but is doing the bug creation.

File: tools/IoTEdgeDevOps/DevOpsLib/BuildDefinitionId.cs [hunk 5]
```diff
@@ -11,7 +11,7 @@ public enum BuildDefinitionId
         EndToEndTest = 87020,
         NestedEndToEndTest = 183241,
         ImageRelease = 31987,
-        LibiohsmCI = 39853,
+        LibiothsmCI = 39853,
```
Comment (and-rewsmith) [inline tools/IoTEdgeDevOps/DevOpsLib/BuildDefinitionId.cs:14]: Fixed this typo.

File: tools/IoTEdgeDevOps/DevOpsLib/BugManagement.cs (hunk 4 above)
Comment (nlcamp) [reply tools/IoTEdgeDevOps/DevOpsLib/BugManagement.cs:2]: Are both files needed? Or would it be better/simpler to consolidate them into a single BugManagement file and class? I see that they both use different subsets of the overarching Work Item Tracking APIs and they both perform operations related to work items: One creates bugs and the other gets the bug count. It seems reasonable to have both of these operations fall under the same abstraction.

Comment (and-rewsmith) [reply tools/IoTEdgeDevOps/DevOpsLib/BugManagement.cs:2]: The pattern is that each individual api has its own class. These apis share nothing other than the partial name "Bug", so I thought it best to keep following the pattern and have separate classes.

File: tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs [hunk 6]
```diff
@@ -149,6 +205,77 @@ void UpsertVstsBugToDb(SqlConnection sqlConnection, BugQuery bugQuery, int bugCo
...
+            finally
+            {
+                sqlConnection?.Close();
+            }
+
+            return filteredBuilds;
+        }
+
+        // We don't want to spam making bugs.
+        // Given this logic runs at least every few minutes, only make a bug for builds in the past 1 hour.
```
Comment (nlcamp) [inline tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs:255]: What causes this logic to run every few minutes? Is it scheduled? Or is it triggered each time a build completes?

**Review by nlcamp — APPROVED** [review]
This is great!

File: tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs (hunk 6 above)
Comment (and-rewsmith) [reply tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs:255]: This loop:
https://github.com/and-rewsmith/iotedge/blob/andsmi/vsts-sync-bug-automation/tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs#L41

**Pull request merged**